File: DSCI553/hw4/task1.py
import sys
from pyspark import SparkContext
from pyspark.sql import SQLContext
from graphframes import GraphFrame

def main():
    filter_threshold = int(sys.argv[1])
    input_file_path = sys.argv[2]
    community_output_file_path = sys.argv[3]
    
    sc = SparkContext('local[*]', 'task1')
    sc.setLogLevel('ERROR')
    sqlContext = SQLContext(sc)
    
    rdd = sc.textFile(input_file_path)
    header = rdd.first()
    data = rdd.filter(lambda line: line != header).map(lambda line: line.split(','))
    
    user_business = data.map(lambda x: (x[0], x[1]))
    
    user_businesses = user_business.groupByKey().mapValues(set).collectAsMap()
    
    all_users = set(user_businesses.keys())
    
    edges = []
    users_list = sorted(all_users)
    
    for i in range(len(users_list)):
        for j in range(i + 1, len(users_list)):
            user1 = users_list[i]
            user2 = users_list[j]
            
            if user1 in user_businesses and user2 in user_businesses:
                common_businesses = len(user_businesses[user1] & user_businesses[user2])
                
                if common_businesses >= filter_threshold:
                    edges.append((user1, user2))
                    edges.append((user2, user1))
    
    nodes_with_edges = set()
    for edge in edges:
        nodes_with_edges.add(edge[0])
        nodes_with_edges.add(edge[1])
    
    vertices = sqlContext.createDataFrame([(user,) for user in nodes_with_edges], ['id'])
    edges_df = sqlContext.createDataFrame(edges, ['src', 'dst'])
    
    graph = GraphFrame(vertices, edges_df)
    
    result = graph.labelPropagation(maxIter=5)
    
    communities_rdd = result.rdd.map(lambda row: (row['label'], row['id']))
    communities = communities_rdd.groupByKey().mapValues(list).map(lambda x: sorted(x[1])).collect()
    
    communities.sort(key=lambda x: (len(x), x[0]))
    
    with open(community_output_file_path, 'w') as f:
        for community in communities:
            f.write("'" + "', '".join(community) + "'\n")
    
    sc.stop()

if __name__ == '__main__':
    main()

File: DSCI553/hw4/src/main/scala/task1.scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.SparkContext
import org.apache.spark.sql.functions._
import org.graphframes.GraphFrame

object task1 {
  def main(args: Array[String]): Unit = {
    val filterThreshold = args(0).toInt
    val inputFilePath = args(1)
    val communityOutputPath = args(2)
    
    val spark = SparkSession
      .builder()
      .appName("task1")
      .master("local[*]")
      .getOrCreate()
    
    val sc = spark.sparkContext
    sc.setLogLevel("ERROR")
    
    import spark.implicits._
    
    val data = sc.textFile(inputFilePath)
    val header = data.first()
    val records = data.filter(line => line != header)
      .map(line => {
        val parts = line.split(",")
        (parts(0), parts(1))
      })
    
    val userBusinesses = records.groupByKey()
      .mapValues(_.toSet)
      .collectAsMap()
    
    val allUsers = userBusinesses.keys.toList.sorted
    
    val edges = scala.collection.mutable.ArrayBuffer[(String, String)]()
    
    for (i <- allUsers.indices) {
      for (j <- i + 1 until allUsers.length) {
        val user1 = allUsers(i)
        val user2 = allUsers(j)
        
        if (userBusinesses.contains(user1) && userBusinesses.contains(user2)) {
          val commonBusinesses = userBusinesses(user1).intersect(userBusinesses(user2))
          
          if (commonBusinesses.size >= filterThreshold) {
            edges += ((user1, user2))
            edges += ((user2, user1))
          }
        }
      }
    }
    
    val nodesWithEdges = edges.flatMap(e => List(e._1, e._2)).toSet
    
    val vertices = spark.createDataFrame(
      nodesWithEdges.toSeq.map(Tuple1(_))
    ).toDF("id")
    
    val edgesDF = spark.createDataFrame(edges).toDF("src", "dst")
    
    val graph = GraphFrame(vertices, edgesDF)
    
    val result = graph.labelPropagation.maxIter(5).run()
    
    val communities = result.rdd
      .map(row => (row.getAs[Long]("label"), row.getAs[String]("id")))
      .groupByKey()
      .map { case (_, users) => users.toList.sorted }
      .collect()
    
    val sortedCommunities = communities.sortBy(c => (c.size, c.head))
    
    val writer = new java.io.PrintWriter(communityOutputPath)
    try {
      for (community <- sortedCommunities) {
        writer.println(community.map(u => s"'$u'").mkString(", "))
      }
    } finally {
      writer.close()
    }
    
    spark.stop()
  }
}